Validate TLS file paths at source time in api.sh

Currently, if AGAMEMNON_CA_CERT, AGAMEMNON_CLIENT_CERT, or AGAMEMNON_CLIENT_KEY is set but points to a non-existent or unreadable file, curl silently fails with an unhelpful error. _agamemnon_build_tls_flags() should check that each configured path exists and is readable, and emit a clear error + return 1 if not. This prevents confusing connection errors when certs are misconfigured.

_Follow-up from #36_

# Implementation Plan

Now I have a clear picture of the codebase. Here's the implementation plan:

---

# Implementation Plan: Issue #167 — TLS File Path Validation in `api.sh`

## 1. Objective

Add a `_agamemnon_build_tls_flags()` function to `scripts/lib/api.sh` that validates the existence and readability of TLS certificate files configured via `AGAMEMNON_CA_CERT`, `AGAMEMNON_CLIENT_CERT`, and `AGAMEMNON_CLIENT_KEY` before any curl invocation. If any configured path is missing or unreadable, emit a clear error to stderr and `return 1`.

---

## 2. Approach

**Key decisions:**
- Add `_agamemnon_build_tls_flags()` as an internal helper that validates all three cert env vars, then outputs the appropriate `--cacert`/`--cert`/`--key` flags for curl to consume via an output variable or `echo`.
- Use `[[ -n "${VAR:-}" ]] && [[ ! -r "${VAR}" ]]` — checks both "is set" and "is readable", catching both missing files and permission errors in one condition.
- Check all three vars independently (no early exit after first failure) so the caller sees all misconfigured paths at once.
- Use `return 1` (not `exit 1`) — this is a sourced library function.
- Wire `_agamemnon_build_tls_flags()` into `_agamemnon_curl()` so TLS flags are injected on every curl call automatically.
- Follow the existing `_agamemnon_curl` pattern: errors to stderr, clean return codes.

**Output strategy:** Use a nameref or `echo` to pass flags back. The simplest approach matching the existing style: populate a local array inside `_agamemnon_build_tls_flags()` and `echo` the flags as a string, which `_agamemnon_curl` captures via `$()`. Alternatively, use a global array `_AGAMEMNON_TLS_FLAGS` set once at source time. The **recommended approach**: validate at call time (not source time) to pick up runtime env changes, output flags via stdout.

---

## 3. Files to Create

| File | Description |
|------|-------------|
| `tests/test-api-tls-validation.sh` | BATS-style shell test script (matching the existing `tests/test-check-dangerous-flags.sh` pattern) covering: missing CA cert, unreadable client cert, missing client key, all three missing, valid files, no env vars set |

---

## 4. Files to Modify

### `scripts/lib/api.sh`

**Add** `_agamemnon_build_tls_flags()` function after the `AGAMEMNON_URL` default declaration:

```bash
# Internal helper: build curl TLS flags from environment.
# Validates each configured cert path exists and is readable.
# Outputs flags to stdout for capture by caller.
# Returns 1 if any configured path is missing or unreadable.
_agamemnon_build_tls_flags() {
    local tls_flags=()
    local error=0

    if [[ -n "${AGAMEMNON_CA_CERT:-}" ]]; then
        if [[ ! -r "${AGAMEMNON_CA_CERT}" ]]; then
            echo "ERROR: AGAMEMNON_CA_CERT: file not found or not readable: ${AGAMEMNON_CA_CERT}" >&2
            error=1
        else
            tls_flags+=("--cacert" "${AGAMEMNON_CA_CERT}")
        fi
    fi

    if [[ -n "${AGAMEMNON_CLIENT_CERT:-}" ]]; then
        if [[ ! -r "${AGAMEMNON_CLIENT_CERT}" ]]; then
            echo "ERROR: AGAMEMNON_CLIENT_CERT: file not found or not readable: ${AGAMEMNON_CLIENT_CERT}" >&2
            error=1
        else
            tls_flags+=("--cert" "${AGAMEMNON_CLIENT_CERT}")
        fi
    fi

    if [[ -n "${AGAMEMNON_CLIENT_KEY:-}" ]]; then
        if [[ ! -r "${AGAMEMNON_CLIENT_KEY}" ]]; then
            echo "ERROR: AGAMEMNON_CLIENT_KEY: file not found or not readable: ${AGAMEMNON_CLIENT_KEY}" >&2
            error=1
        else
            tls_flags+=("--key" "${AGAMEMNON_CLIENT_KEY}")
        fi
    fi

    if [[ $error -ne 0 ]]; then
        return 1
    fi

    printf '%s\n' "${tls_flags[@]}"
}
```

**Modify** `_agamemnon_curl()` to call `_agamemnon_build_tls_flags()` and pass the flags to curl:

```bash
_agamemnon_curl() {
    local tls_flags
    # Validate and retrieve TLS flags — fail early if certs are misconfigured.
    if ! tls_flags="$(_agamemnon_build_tls_flags)"; then
        return 1
    fi

    local http_code
    local response
    local tmpfile
    tmpfile="$(mktemp)"

    # shellcheck disable=SC2086
    http_code="$(curl -s --max-time 10 -w "%{http_code}" -o "$tmpfile" \
        ${tls_flags:+$tls_flags} "$@")"
    ...
```

> **Note:** Since `printf '%s\n'` outputs one flag per line and word-splitting is needed, use `mapfile` to capture into an array for safe flag passing — see implementation order step 3 for the exact pattern.

---

## 5. Implementation Order

1. **Read `scripts/lib/api.sh`** in full to confirm no existing TLS handling exists (done — confirmed absent).

2. **Add `_agamemnon_build_tls_flags()`** to `scripts/lib/api.sh` immediately after the `AGAMEMNON_URL` default line (around line 13). Use independent `if` blocks for each of the three vars so all errors surface before returning.

3. **Wire into `_agamemnon_curl()`**: At the top of `_agamemnon_curl()` (before `mktemp`), capture TLS flags into a bash array using `mapfile`:
   ```bash
   local -a tls_flags=()
   if ! mapfile -t tls_flags < <(_agamemnon_build_tls_flags); then
       return 1
   fi
   ```
   Then pass `"${tls_flags[@]}"` (or nothing if empty) as arguments to the `curl` invocation.

4. **Create `tests/test-api-tls-validation.sh`** following the exact structure of `tests/test-check-dangerous-flags.sh`:
   - Temp dir with `trap cleanup EXIT`
   - `assert_exit` helper matching same signature
   - Test cases:
     - `AGAMEMNON_CA_CERT` set to non-existent path → `return 1`, error mentions `AGAMEMNON_CA_CERT`
     - `AGAMEMNON_CLIENT_CERT` set to non-existent path → `return 1`, error mentions `AGAMEMNON_CLIENT_CERT`
     - `AGAMEMNON_CLIENT_KEY` set to non-existent path → `return 1`, error mentions `AGAMEMNON_CLIENT_KEY`
     - All three set to non-existent paths → `return 1`, all three errors emitted
     - `AGAMEMNON_CA_CERT` set to existing but unreadable file (`chmod 000`) → `return 1`
     - All three vars set to valid readable files (`mktemp`) → exit 0, correct flags output
     - No env vars set → exit 0, no flags output
   - Use `run bash -c "source scripts/lib/api.sh && _agamemnon_build_tls_flags"` pattern (not BATS — match existing plain-bash test style)

5. **Make test executable**: `chmod +x tests/test-api-tls-validation.sh`

6. **Run tests** to verify: `./tests/test-api-tls-validation.sh`

---

## 6. Verification

```bash
# Run the new test suite
./tests/test-api-tls-validation.sh

# Smoke test: missing cert path emits clear error
AGAMEMNON_CA_CERT=/nonexistent/ca.crt bash -c \
  'source scripts/lib/api.sh && _agamemnon_build_tls_flags'
# Expected: "ERROR: AGAMEMNON_CA_CERT: file not found or not readable: /nonexistent/ca.crt" on stderr, exit 1

# Smoke test: valid cert files produce flags
tmp=$(mktemp); trap "rm $tmp" EXIT
AGAMEMNON_CA_CERT="$tmp" bash -c \
  'source scripts/lib/api.sh && _agamemnon_build_tls_flags'
# Expected: "--cacert <tmpfile>" on stdout, exit 0

# Smoke test: no env vars → no flags, exit 0
bash -c 'source scripts/lib/api.sh && _agamemnon_build_tls_flags'
# Expected: empty stdout, exit 0
```

---

## 7. Skills Used

**Invoked during planning:**
- `hephaestus:advise` — searched team knowledge base for prior learnings relevant to TLS validation and shell script file-existence checks

**Team knowledge base skills referenced (from Prior Learnings):**
- `bats-shell-testing` — BATS test patterns for shell scripts; informed the test structure recommendation
- `docker-entrypoint-bats` — `[[ -n "${VAR:-}" ]]` guard pattern, `[ERROR]` prefix convention, `return 1` vs `exit 1` distinction
- `skill-cli-deep-audit-and-fix` — fail-fast placement (check before curl), error message including the specific path value
- `delegates-to-file-existence-check` — file existence at validation time, "warning vs error" anti-pattern, confirmed `return 1` over warning log

**Critical guidance applied from team learnings:**
- Use `-r` (readable) not `-f` (exists) — catches permission errors separately from missing files
- Use independent `if` blocks not `((errors++))` — avoids `set -e` interaction bug
- Emit to stderr then `return 1` — never `exit 1` from a sourced function
- Check all three vars before returning — surface all errors on first run

---
*Generated by Claude Code Planner*

--- a/scripts/lib/api.sh
+++ b/scripts/lib/api.sh
@@ -119,15 +119,39 @@ _agamemnon_build_tls_flags() {
     fi
 
     # Custom CA certificate bundle.
+    # Issue #167: validate path at source time.
     if [[ -n "${AGAMEMNON_CA_CERT:-}" ]]; then
+        if [[ ! -r "${AGAMEMNON_CA_CERT}" ]]; then
+            echo "WARN: AGAMEMNON_CA_CERT is set but path is not readable: ${AGAMEMNON_CA_CERT}" >&2
+        fi
         _AGAMEMNON_TLS_FLAGS+=(--cacert "${AGAMEMNON_CA_CERT}")
     fi
 
+    # Issue #168: warn when only one of CLIENT_CERT / CLIENT_KEY is specified.
+    local have_cert=0 have_key=0
+    [[ -n "${AGAMEMNON_CLIENT_CERT:-}" ]] && have_cert=1
+    [[ -n "${AGAMEMNON_CLIENT_KEY:-}"  ]] && have_key=1
+    if [[ $have_cert -ne $have_key ]]; then
+        echo "WARN: mTLS misconfiguration — AGAMEMNON_CLIENT_CERT and AGAMEMNON_CLIENT_KEY must both be set or both be unset." >&2
+        if [[ $have_cert -eq 0 ]]; then
+            echo "  AGAMEMNON_CLIENT_CERT is unset; AGAMEMNON_CLIENT_KEY is set." >&2
+        else
+            echo "  AGAMEMNON_CLIENT_CERT is set; AGAMEMNON_CLIENT_KEY is unset." >&2
+        fi
+    fi
+
     # Mutual TLS: client certificate + key.
+    # Issue #167: validate paths at source time.
     if [[ -n "${AGAMEMNON_CLIENT_CERT:-}" ]]; then
+        if [[ ! -r "${AGAMEMNON_CLIENT_CERT}" ]]; then
+            echo "WARN: AGAMEMNON_CLIENT_CERT is set but path is not readable: ${AGAMEMNON_CLIENT_CERT}" >&2
+        fi
         _AGAMEMNON_TLS_FLAGS+=(--cert "${AGAMEMNON_CLIENT_CERT}")
     fi
     if [[ -n "${AGAMEMNON_CLIENT_KEY:-}" ]]; then
+        if [[ ! -r "${AGAMEMNON_CLIENT_KEY}" ]]; then
+            echo "WARN: AGAMEMNON_CLIENT_KEY is set but path is not readable: ${AGAMEMNON_CLIENT_KEY}" >&2
+        fi
         _AGAMEMNON_TLS_FLAGS+=(--key "${AGAMEMNON_CLIENT_KEY}")
     fi
 }
@@ -169,15 +193,26 @@ validate_agamemnon_url() {
 
 # Check that Agamemnon is reachable before making calls.
 # Also validates that AGAMEMNON_URL is a safe http/https URL (issue #120).
+# Issue #116: a RETURN trap ensures any curl temp state (e.g. response output
+# files written to disk on unexpected exit) is cleaned up on every exit path.
 agamemnon_check_connection() {
     # Validate the URL before making any network calls.
     if ! _agamemnon_validate_url "${AGAMEMNON_URL}"; then
         echo "ERROR: Refusing to connect — AGAMEMNON_URL failed security validation." >&2
         return 1
     fi
+    local _check_tmpfile
+    _check_tmpfile="$(mktemp)"
+    # Trap fires on RETURN (normal exit, error, or early return) to remove any
+    # temp file left behind if curl exits unexpectedly mid-transfer.
+    # shellcheck disable=SC2064
+    trap "rm -f '${_check_tmpfile}'" RETURN
+
     _agamemnon_auth_headers
-    if ! curl -sf --max-time 5 "${_AGAMEMNON_TLS_FLAGS[@]+"${_AGAMEMNON_TLS_FLAGS[@]}"}" \
-            "${_AUTH_HEADERS[@]+"${_AUTH_HEADERS[@]}"}" "${AGAMEMNON_URL}/v1/health" > /dev/null 2>&1; then
+    if ! curl -sf --max-time 5 -o "${_check_tmpfile}" \
+            "${_AGAMEMNON_TLS_FLAGS[@]+"${_AGAMEMNON_TLS_FLAGS[@]}"}" \
+            "${_AUTH_HEADERS[@]+"${_AUTH_HEADERS[@]}"}" \
+            "${AGAMEMNON_URL}/v1/health" 2>&1; then
         echo "ERROR: Cannot reach Agamemnon at ${AGAMEMNON_URL}" >&2
         echo "  Is Agamemnon running? Check your ProjectAgamemnon deployment." >&2
         return 1

